vm/src/stdlib/ctypes/array.rs

```diff
@@ -0,0 +1,633 @@
+    BorrowValueMut, PyCData, PyCDataFunctions, PyCDataMethods, PyCDataSequenceMethods, RawBuffer,
+};
+use crate::stdlib::ctypes::pointer::PyCPointer;
+use crate::stdlib::ctypes::primitive::{new_simple_type, PySimpleType};
+
+macro_rules! byte_match_type {
+    (
+        $kind: expr,
+        $byte: expr,
+        $vm: expr,
+        $(
+            $($type: literal)|+ => $body: ident
+        )+
+    ) => {
+        match $kind {
+            $(
+                $(
+                    t if t == $type => {
+                        let chunk: [u8; mem::size_of::<$body>()] = $byte.try_into().unwrap();
+                        $vm.new_pyobj($body::from_ne_bytes(chunk))
+                    }
+                )+
+            )+
+            _ => unreachable!()
+        }
+    }
+}
+
+fn byte_to_pyobj(ty: &str, b: &[u8], vm: &VirtualMachine) -> PyObjectRef {
+    if ty == "u" {
+        vm.new_pyobj(if cfg!(windows) {
+            let chunk: [u8; 2] = b.try_into().unwrap();
+            u16::from_ne_bytes(chunk) as u32
+        } else {
+            let chunk: [u8; 4] = b.try_into().unwrap();
+            u32::from_ne_bytes(chunk)
+        })
+    } else {
+        byte_match_type!(
+            ty,
+            b,
+            vm,
+            "c" | "b" => i8
+            "h" => i16
+            "H" => u16
+            "i" => i32
+            "I" => u32
+            "l" | "q" => i64
+            "L" | "Q" => u64
+            "f" => f32
+            "d" | "g" => f64
+            "?" | "B" => u8
+            "P" | "z" | "Z" => usize
+        )
+    }
+}
+
+fn slice_adjust_size(length: isize, start: &mut isize, stop: &mut isize, step: isize) -> isize {
+    if *start < 0 {
+        *start += length;
+        if *start < 0 {
+            *start = if step < 0 { -1 } else { 0 };
+        }
+    } else if *start >= length {
+        *start = if step < 0 { length - 1 } else { length };
+    }
+
+    if *stop < 0 {
+        *stop += length;
+        if *stop < 0 {
+            *stop = if step < 0 { -1 } else { 0 };
+        }
+    } else if *stop >= length {
+        *stop = if step < 0 { length - 1 } else { length };
+    }
+
+    if step < 0 {
+        if *stop < *start {
+            return (*start - *stop - 1) / (-step) + 1;
+        }
+    } else if *start < *stop {
+        return (*stop - *start - 1) / step + 1;
+    }
+
+    0
+}
+
+pub fn make_array_with_lenght(
+    cls: PyTypeRef,
+    length: usize,
+    vm: &VirtualMachine,
+) -> PyResult<PyRef<PyCArray>> {
+    if let Some(outer_type) = cls.get_attr("_type_") {
+        let length = length as usize;
+        if let Ok(_type_) = vm.get_attribute(outer_type.clone(), "_type_") {
+            let itemsize = get_size(_type_.downcast::<PyStr>().unwrap().to_string().as_str());
+
+            if length.checked_mul(itemsize).is_none() {
+                Err(vm.new_overflow_error("Array size too big".to_string()))
+            } else {
```

Comment:
```suggestion
            length.checked_mul(itemsize).unwrap_or_else(|| Err(vm.new_overflow_error("Array size too big".to_owned()))?;
```

then the whole block will use one-level less indent